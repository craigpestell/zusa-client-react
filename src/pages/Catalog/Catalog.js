/* @flow */

import React, { PureComponent } from 'react';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import Helmet from 'react-helmet';
import { catalogAction } from '../../actions';
import type { Catalog as CatalogType, Dispatch, ReduxState } from '../../types';
import { CatalogList } from '../../components';
import styles from './styles.scss';

type Props = { catalog: CatalogType, fetchCatalogIfNeeded: () => void };

// Export this for unit testing more easily
export class Catalog extends PureComponent<Props> {
  componentDidMount() {
    const { fetchCatalogIfNeeded } = this.props;

    fetchCatalogIfNeeded();
  }

  renderCatalogList = () => {
    const { catalog } = this.props;
    if (
      !catalog.readyStatus ||
      catalog.readyStatus === 'CATALOG_INVALID' ||
      catalog.readyStatus === 'CATALOG_REQUESTING'
    ) {
      return <p>Loading...</p>;
    }
    if (catalog.readyStatus === 'CATALOG_FAILURE') {
      return <p>Oops, Failed to load list!</p>;
    }

    return (
      <div className={styles.Catalog}>
        <Helmet title="Catalog" />
        <CatalogList list={catalog.list.data.workOrders} />
      </div>
    );
  };

  render() {
    return <div className={styles.Catalog}>{this.renderCatalogList()}</div>;
  }
}

const connector = connect(
  ({ catalog }: ReduxState) => ({ catalog }),
  (dispatch: Dispatch) => ({
    fetchCatalogIfNeeded: () => dispatch(catalogAction.fetchCatalogIfNeeded())
  })
);

export default compose(
  withRouter,
  connector
)(Catalog);
