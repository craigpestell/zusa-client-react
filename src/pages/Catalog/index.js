/* @flow */

import loadable from 'loadable-components';

import { ErrorDisplay, Loading } from '../../components';

export default loadable(() => import('./Catalog'), {
  ErrorComponent: ErrorDisplay,
  LoadingComponent: Loading
});
