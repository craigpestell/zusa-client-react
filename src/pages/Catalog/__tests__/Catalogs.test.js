import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';

import { Catalog } from '../Catalog';

describe('<Catalog />', () => {
  const tree = (props, actions) =>
    renderer
      .create(
        <MemoryRouter>
          <Catalog {...props} {...actions} />
        </MemoryRouter>
      )
      .toJSON();

  test('should call fetchCatalogIfNeeded when componentDidMount', () => {
    const mockAction = jest.fn();
    const props = {
      catalog: {}
    };
    const actions = {
      fetchCatalogIfNeeded: mockAction
    };

    mount(
      <MemoryRouter>
        <Catalog {...props} {...actions} />
      </MemoryRouter>
    );

    expect(mockAction).toHaveBeenCalled();
  });

  test('renders the loading status if data invalid', () => {
    const props = {
      catalog: { readyStatus: 'CATALOG_INVALID' }
    };
    const actions = { fetchCatalogIfNeeded: () => {} };

    expect(tree(props, actions)).toMatchSnapshot();
  });

  test('renders the loading status if requesting data', () => {
    const props = {
      catalog: { readyStatus: 'CATALOG_REQUESTING' }
    };
    const actions = { fetchCatalogIfNeeded: () => {} };

    expect(tree(props, actions)).toMatchSnapshot();
  });

  test('renders an error if loading failed', () => {
    const props = {
      catalog: { readyStatus: 'CATALOG_FAILURE' }
    };
    const actions = { fetchCatalogIfNeeded: () => {} };

    expect(tree(props, actions)).toMatchSnapshot();
  });

  test('renders the <CatalogList /> if loading was successful', () => {
    const props = {
      catalog: {
        readyStatus: 'CATALOG_SUCCESS',
        list: [{ id: '1', name: 'Welly' }]
      }
    };
    const actions = { fetchCatalogIfNeeded: () => {} };

    expect(tree(props, actions)).toMatchSnapshot();
  });
});
