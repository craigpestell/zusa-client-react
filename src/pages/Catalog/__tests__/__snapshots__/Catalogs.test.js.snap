// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Catalog /> renders an error if loading failed 1`] = `
<div
  className="Catalog"
>
  <p>
    Oops, Failed to load list!
  </p>
</div>
`;

exports[`<Catalog /> renders the <CatalogList /> if loading was successful 1`] = `
<div
  className="Catalog"
>
  <div
    className="CatalogList"
  >
    <h4>
      Catalog List
    </h4>
    <ul>
      <li>
        <a
          href="/wo/1"
          onClick={[Function]}
        >
          Welly
        </a>
      </li>
    </ul>
  </div>
</div>
`;

exports[`<Catalog /> renders the loading status if data invalid 1`] = `
<div
  className="Catalog"
>
  <p>
    Loading...
  </p>
</div>
`;

exports[`<Catalog /> renders the loading status if requesting data 1`] = `
<div
  className="Catalog"
>
  <p>
    Loading...
  </p>
</div>
`;
